src/coreclr/nativeaot/Runtime/eventpipe/ep-rt-aot.h

```diff
@@ -1386,22 +1390,23 @@ ep_rt_utf8_to_utf16le_string (
     if (!str)
         return NULL;
 
-    // Shipping criteria: no EVENTPIPE-NATIVEAOT-TODO left in the codebase
-    // Implementation would just use strlen and malloc to make a new buffer, and would then copy the string chars one by one.
-    // Assumes that only ASCII is used for ep_char8_t
-    size_t len_utf8 = strlen(str);
-    ep_char16_t *str_utf16 = reinterpret_cast<ep_char16_t *>(malloc ((len_utf8 + 1) * sizeof (ep_char16_t)));
-    if (!str_utf16)
+    int32_t flags = MINIPAL_MB_NO_REPLACE_INVALID_CHARS;
+
+    ep_char16_t* lpDestStr = NULL;
+
+    if (static_cast<int>(len) < 0)
+        len = strlen(str) + 1;
```

Comment:
Why do we need the +1 here? Adding the null terminator is handled explicitly below?

Reply:
This is an existing pattern:
https://github.com/dotnet/runtime/blob/8f8a0074ae7287686bcc4b548790ec7ffa565064/src/mono/mono/eglib/giconv.c#L340-L343

Reason is, minipal get_length API needs the source length with null-terminator accounted for, so it returns the correct destination count. Handling this it within minipal was requiring some additional conditions which we decided to leave out for the caller to handle.

Reply:
Does the length in `ep_rt_utf16_to_utf8_string` should need +1 too, then?

Reply:
> This is an existing pattern:

It has the same issue of allocating one extra byte than necessary, and it is further complicated by returning `items_written`.

We do not need to replicate the issue here.

Reply:
@jkotas, not sure which issue you are referring to. The easiest way to understand what's goes wrong without +1 is to remove it from coreclr version (it's on the callsite of PAL API) and run PAL tests. https://github.com/dotnet/runtime/commit/fb613a56f05d6da66a6e16efa9319f0000b43540 did the same thing.

Reply:
@jkotas, not sure which issue you are referring to. The easiest way to understand what goes wrong without +1 is to remove it from coreclr version (it's on the callsite of PAL API) and run PAL tests. https://github.com/dotnet/runtime/commit/fb613a56f05d6da66a6e16efa9319f0000b43540 did the same thing.

Reply:
> @jkotas, not sure which issue you are referring to.

Allocating buffer that is one character longer than necessary.

For example, if `ep_rt_utf8_to_utf16le_string` is called with a string that is 1 ascii character + zero terminator, the current code allocates `3 * sizeof(CHAR16_T)`. It is one `CHAR16_T` more than what is actually needed.

Reply:
It seems like the pattern with the extra +1 that @am11 mentions is needed to preserve the current behavior. For example, this [test ](https://github.com/dotnet/runtime/blob/main/src/tests/tracing/eventpipe/providervalidation/providervalidation.cs)fails without +1.

Reply:
Are you saying that the event pipe code expects the strings returned by this method to have double zero terminators? It is very atypical contract. If it is really the case, these should be a long comment explaining it.

Reply:
> Allocating buffer that is one character longer than necessary.

PAL, COM and reflection tests are very sensitive about "fill the exact slot or fail" cases. So if that was the case, they all will fail. IOW, that +1 effect gets cancelled out how the code in utf8.c works.